torchrl/objectives/returns/functional.py

```diff
@@ -195,9 +196,31 @@ def _custom_conv1d(tensor, filter):
         -1,
     )
 
-    # shape = val.shape
-    filter = filter.squeeze(-1).unsqueeze(0).unsqueeze(0)  # 1 x 1 x T
-    out = torch.conv1d(val_pad, filter)
+    if filter.ndimension() > 2:
+        # filter will have shape batch_dims x timesteps x timesteps x 1
+        # reshape to batch_dims x timesteps x 1 x 1 x timesteps ready for convolving
+        filter = filter.reshape(*filter.shape[:-2], 1, 1, filter.shape[-2])
+
+        # add dimension to val-pad as we are going to vectorise over the first
+        # axis (batch) since we have different filter per batch
+        val_pad = val_pad.unsqueeze(1)
+
+        # vectorise application of filters to each batch. we have multiple
+        # filters and just one trajectory to apply it to.
+        # TODO: there's a lot of redundancy here since we only care about the
+        # diagonal entries. how to simplify?
+        conv1d_batch = functorch.vmap(torch.conv1d, (None, 0))
+
+        # vectorise over batch dimension
+        out = functorch.vmap(conv1d_batch)(val_pad, filter)
+
+        # out has dimensions batch x timesteps x 1 x 1 x timesteps
+        # we only care about the diagonal timestep entries
+        out = torch.diagonal(out, dim1=-4, dim2=-1).squeeze(-3)
```

Comment:
This section, which I've tried to add plenty of comments to, is where I would particularly like some guidance as to whether I have got things right. Even if this is right, it seems like there's some redundancy which could likely be made more efficient

Suppose this has been called from line 314 `rs = _custom_conv1d(reward, gammas * lambdas)`. Thus `tensor` has dimensions `batch x 1 x timesteps` and `filter` has dimensions `batch x timesteps x timesteps x 1` (before reshaping). There are two dimensions length `timesteps` because we have one `gamma` per batch and timestep (call it $\gamma_{bt}$), and each of those we calculate the decay sequence $\gamma_{bt}, \gamma^2_{bt}, \dots$.

The return value has the same dimensions as `tensor` (`batch x 1 x timesteps`). My understanding is that it should contain values like

$$
    \text{out}[b, 0, t] = r_t + \lambda \gamma_{bt} r_{t+1} + \lambda^2\gamma^2_{bt}r_{t+2} + \dots
$$

These can still be calculated using a convolution, but notice that because `t` appears both in the gamma index and the offset of the rewards, we only actually care about the diagonal values of the full vectorised convolution.

Does that seem right? If so I did think for a while about whether it could be done more directly but didn't solve it yet.

If it doesn't seem right, pointers are appreciated!

Reply:
This runs for me and gives the expected shape. However I'm a bit unclear on the inputs. We are allowing the user to specify a `gamma` per batch _and_ per timestep right? So would that not imply that `filter` should have dimensions for batch and timesteps (corresponding to the different `gamma`s) _and_ another dimension for the geometric decay over subsequent timesteps? Or have I misunderstood that?

The above is the assumption I made in my first pass, which means I'm not sure how to compare my implementation to your snippet?

Reply:
This discussion no longer relevant after switch to `einsum` which takes care of the redundancy.